### ISSUE TITLE
tf.data.Iterator.from_string_handle() breaking behaviour in r1.4 compared to r1.3.1

### ISSUE THREAD

**Issue by MtDersvan** [tensorflow/tensorflow#14613]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: +
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.4.0-rc1-11-g130a514 1.4.0
- **Python version**: Python 3.5.2
- **Bazel version (if compiling from source)**: -
- **GCC/Compiler version (if compiling from source)**: -
- **CUDA/cuDNN version**: nvidia/cuda:8.0-cudnn6-devel-ubuntu16.04
- **GPU model and memory**: NVIDIA® Tesla® K80 (GCE)
- **Exact command to reproduce**:

### Context:
Same setup/context as in the previous issue https://github.com/tensorflow/tensorflow/issues/12859 including this fix https://github.com/tensorflow/tensorflow/issues/12859#issuecomment-327783827.
In addition the training is now in a Multi-task Learning mode, so feedable contrib Iterator was used to accommodate different datasets for each task.
Simplified snippet:
~~~python
        self.handle = tf.placeholder(tf.string, shape=[])

        ...

        # targets -> Multi-task training targets.
        # datasets -> dict of Multi-task target tf.Datasets.
        for target in self.targets:
            self.datasets[target] = self.create_TFDataset()

        self.iterator = tf.contrib.data.Iterator.from_string_handle(
            self.handle,
            self.datasets[targets[-1]].output_types,
            self.datasets[targets[-1]].output_shapes)

        # iter_init_ops and iter_handles -> init_ops & handles per each task.
        for target in self.targets:
            self.iterators[target] = self.datasets[target].make_initializable_iterator()
            self.iter_init_ops[target] = self.iterators[target].initializer
            self.iter_handles[target] = self.iterators[target].string_handle()

        ...
        # Within tf.train.MonitoredTrainingSession as mon_sess.
        for target in targets:
            # Get all target datasets handles.
            handle[target] = mon_sess._coordinated_creator.tf_sess.run(
                training_model.iter_handles[target])
            # Init all target datasets.
            mon_sess._coordinated_creator.tf_sess.run(training_model.iter_init_ops[target])

        ...
        # Training step for a specific target.
        input_feed = {self.handle:handle[target]}
        output_feed = [
            self.update_ops[target],
            self.losses[target], 
            self.metrics[target],
        ]
        outputs = session.run(output_feed, input_feed, options=options,
                                             run_metadata=run_metadata)
~~~
### Problem:
This system worked flawlessly for tf 1.3 & tf 1.3.1. 
After a planned update this week to tf 1.4 the following Error would appear at absolute random (it might appear after 5 seconds or after a few hours of training):
~~~console
...
W tensorflow/core/framework/op_kernel.cc:1192] Unavailable: Endpoint read failed
	 [[Node: Model/Generate_BiRNN/BiRNN_Logic/bidirectional_rnn/fw/carry_w_S543 = _Recv[_start_time=0, client_terminated=false, recv_device="/job:worker/replica:0/task:0/device:GPU:0", send_device="/job:ps/replica:0/task:1/device:CPU:0", send_device_incarnation=-458934800929363750, tensor_name="edge_9_Model/Generate_BiRNN/BiRNN_Logic/bidirectional_rnn/fw/carry_w", tensor_type=DT_FLOAT, _device="/job:worker/replica:0/task:0/device:GPU:0"]()]]
	 [[Node: Training_Graph/Model/TARGET/TARGET_Metrics/Select_G315 = _Recv[client_terminated=false, recv_device="/job:worker/replica:0/task:0/device:CPU:0", send_device="/job:worker/replica:0/task:0/device:GPU:0", send_device_incarnation=3243841103411587778, tensor_name="edge_4692_Training_Graph/Model/TARGET/TARGET_Metrics/Select", tensor_type=DT_DOUBLE, _device="/job:worker/replica:0/task:0/device:CPU:0"]()]]
...
W tensorflow/core/framework/op_kernel.cc:1192] Not found: Resource worker/_3_Training_Graph/Model/Iterator/N10tensorflow12_GLOBAL__N_116IteratorResourceE does not exist.
	 [[Node: Training_Graph/Model/IteratorFromStringHandle = IteratorFromStringHandle[output_shapes=[[?,?,?], [?,?,?], [?], [?,?], [?,?], [?], [?]], output_types=[DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_FLOAT, DT_INT32], _device="/job:worker/replica:0/task:0/device:CPU:0"](_recv_Training_Graph/Model/Placeholder_0)]]
	 [[Node: Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26_G1265 = _Recv[client_terminated=false, recv_device="/job:worker/replica:0/task:0/device:GPU:0", send_device="/job:worker/replica:0/task:0/device:CPU:0", send_device_incarnation=-3380386273340330983, tensor_name="edge_4280_Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26", tensor_type=DT_INT64, _device="/job:worker/replica:0/task:0/device:GPU:0"]()]]
...
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1323, in _do_call
    return fn(*args)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1302, in _run_fn
    status, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/errors_impl.py", line 473, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.NotFoundError: Resource worker/_3_Training_Graph/Model/Iterator/N10tensorflow12_GLOBAL__N_116IteratorResourceE does not exist.
	 [[Node: Training_Graph/Model/IteratorFromStringHandle = IteratorFromStringHandle[output_shapes=[[?,?,?], [?,?,?], [?], [?,?], [?,?], [?], [?]], output_types=[DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_FLOAT, DT_INT32], _device="/job:worker/replica:0/task:0/device:CPU:0"](_recv_Training_Graph/Model/Placeholder_0)]]
	 [[Node: Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26_G1265 = _Recv[client_terminated=false, recv_device="/job:worker/replica:0/task:0/device:GPU:0", send_device="/job:worker/replica:0/task:0/device:CPU:0", send_device_incarnation=-3380386273340330983, tensor_name="edge_4280_Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26", tensor_type=DT_INT64, _device="/job:worker/replica:0/task:0/device:GPU:0"]()]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "dist_train.py", line 684, in <module>
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "dist_train.py", line 640, in main
    create_job()
  File "dist_train.py", line 628, in create_job
    run_worker(server, cluster)
  File "dist_train.py", line 429, in run_worker
    profile=profile
  File "/HARNN.py", line 1576, in step_dist_gpu
    run_metadata=run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/monitored_session.py", line 521, in run
    run_metadata=run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/monitored_session.py", line 892, in run
    run_metadata=run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/monitored_session.py", line 967, in run
    raise six.reraise(*original_exc_info)
  File "/usr/local/lib/python3.5/dist-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/monitored_session.py", line 952, in run
    return self._sess.run(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/monitored_session.py", line 1024, in run
    run_metadata=run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/monitored_session.py", line 827, in run
    return self._sess.run(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 889, in run
    run_metadata_ptr)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1120, in _run
    feed_dict_tensor, options, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1317, in _do_run
    options, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1336, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.NotFoundError: Resource worker/_3_Training_Graph/Model/Iterator/N10tensorflow12_GLOBAL__N_116IteratorResourceE does not exist.
	 [[Node: Training_Graph/Model/IteratorFromStringHandle = IteratorFromStringHandle[output_shapes=[[?,?,?], [?,?,?], [?], [?,?], [?,?], [?], [?]], output_types=[DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_FLOAT, DT_INT32], _device="/job:worker/replica:0/task:0/device:CPU:0"](_recv_Training_Graph/Model/Placeholder_0)]]
	 [[Node: Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26_G1265 = _Recv[client_terminated=false, recv_device="/job:worker/replica:0/task:0/device:GPU:0", send_device="/job:worker/replica:0/task:0/device:CPU:0", send_device_incarnation=-3380386273340330983, tensor_name="edge_4280_Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26", tensor_type=DT_INT64, _device="/job:worker/replica:0/task:0/device:GPU:0"]()]]

Caused by op 'Training_Graph/Model/IteratorFromStringHandle', defined at:
  File "dist_train.py", line 684, in <module>
    tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/platform/app.py", line 48, in run
    _sys.exit(main(_sys.argv[:1] + flags_passthrough))
  File "dist_train.py", line 640, in main
    create_job()
  File "dist_train.py", line 628, in create_job
    run_worker(server, cluster)
  File "dist_train.py", line 272, in run_worker
    training_model.build_graph()
  File "/HARNN.py", line 221, in build_graph
    self._init_dataset()
  File "/HARNN.py", line 329, in _init_dataset
    self.datasets[self.targets[-1]].output_shapes)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/data/ops/iterator_ops.py", line 189, in from_string_handle
    output_shapes=nest.flatten(output_shapes))
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/ops/gen_dataset_ops.py", line 662, in iterator_from_string_handle
    output_types=output_types, output_shapes=output_shapes, name=name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 2956, in create_op
    op_def=op_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 1470, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

NotFoundError (see above for traceback): Resource worker/_3_Training_Graph/Model/Iterator/N10tensorflow12_GLOBAL__N_116IteratorResourceE does not exist.
	 [[Node: Training_Graph/Model/IteratorFromStringHandle = IteratorFromStringHandle[output_shapes=[[?,?,?], [?,?,?], [?], [?,?], [?,?], [?], [?]], output_types=[DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_INT64, DT_FLOAT, DT_INT32], _device="/job:worker/replica:0/task:0/device:CPU:0"](_recv_Training_Graph/Model/Placeholder_0)]]
	 [[Node: Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26_G1265 = _Recv[client_terminated=false, recv_device="/job:worker/replica:0/task:0/device:GPU:0", send_device="/job:worker/replica:0/task:0/device:CPU:0", send_device_incarnation=-3380386273340330983, tensor_name="edge_4280_Training_Graph/Model/TARGET/TARGET_Metrics/DenseToDenseSetOperation_26", tensor_type=DT_INT64, _device="/job:worker/replica:0/task:0/device:GPU:0"]()]]
~~~

### Statement:
What exactly have changed in tf.data.Iterator between versions r1.4 and r1.3.1 that causes such an unpleasant behaviour? What can be done to counteract `NotFoundError`?

**Comment by mrry** [comment]
I don't think there was any change that would cause this error. From the error message though it looks like the handle might have been captured (using `Iterator.to_string_handle()`) in a different session from which it is being used. Since you're using `tf.train.MonitoredTrainingSession`, is it possible that the session has been recreated? If that's the case, you'll need to make sure to recompute the handle—and potentially reinitialize the iterator—in the new session.

A possible workaround, which might or might not work depending on the actual cause of the problem, would be to set a `shared_name` when you create the iterator. I believe that should make the handle string stable across restarts of the session (as long as it's not in a task that could have failed).

**Comment by MtDersvan** [comment]
Actually, @mrry, as I've been training past few days with r1.4, I caught `tf.train.MonitoredTrainingSession` occasionally and silently restarting sessions without raising any errors! 
~~~console
# Few hours of training
...
2017-11-19 02:01:35.079083: I tensorflow/core/distributed_runtime/master_session.cc:1004] 
Start master session fc7b867194458c34 with config: 
gpu_options { allow_growth: true } 
allow_soft_placement: true 
graph_options { optimizer_options { global_jit_level: ON_1 } }
 2017-11-19 02:06:11.963118: I tensorflow/core/kernels/shuffle_dataset_op.cc:110] Filling up shuffle buffer (this may take a while): 23453 of 125000
2017-11-19 02:06:21.963333: I tensorflow/core/kernels/shuffle_dataset_op.cc:110] Filling up shuffle buffer (this may take a while): 47689 of 125000
2017-11-19 02:06:31.963203: I tensorflow/core/kernels/shuffle_dataset_op.cc:110] Filling up shuffle buffer (this may take a while): 70957 of 125000
2017-11-19 02:06:41.963338: I tensorflow/core/kernels/shuffle_dataset_op.cc:110] Filling up shuffle buffer (this may take a while): 95095 of 125000
2017-11-19 02:06:51.963216: I tensorflow/core/kernels/shuffle_dataset_op.cc:110] Filling up shuffle buffer (this may take a while): 119359 of 125000
2017-11-19 02:06:54.333142: I tensorflow/core/kernels/shuffle_dataset_op.cc:121] Shuffle buffer filled. 
...
# Few hours of training
~~~
And this would happen on any worker without any traces at random moments during training.
So I think the Error Handling might have changed since the latest release.
This is not good, and makes my models sad(
Can I be of any help in determining the cause?

**Comment by MtDersvan** [comment]
@mrry @cy89 
I'm getting following behaviour without any warnings or errors, only upper-mentioned `Start master session ...` message. 
![image](https://user-images.githubusercontent.com/7069222/33189189-55b3b19a-d055-11e7-8e61-5361f88354fb.png)
I would like to prevent or at least debug this.

**Comment by cy89** [comment]
@mrry have you seen this kind of behavior before, or might you have any hints?

**Comment by mrry** [comment]
I'm at a loss to explain why the restarts are having such an effect on your training job. I believe `MonitoredTrainingSession` has always attempted to restart the underlying session transparently when certain types of failure happen, and the worker for which `is_chief` is `True` will attempt to reinitialize the parameters from the latest checkpoint if one is available. I doubt the dataset initialization is the cause, although your setup will reinitialize the iterator at each of these restarts, so you might need to shuffle the data to ensure that you're not iterating over the data in the same order each time and thus overfitting somehow.

Can you share more details about the training setup?

/cc @ispirmustafa in case there were any recent changes to `MonitoredSession` that could explain a change in behavior.

**Comment by ispirmustafa** [comment]
I'm too not aware of any change to MonitoredSession that is related. 
BTW, following code segment accesses to private parts of MonitoredSession which may work in an unexpected way:
```
        for target in targets:
            # Get all target datasets handles.
            handle[target] = mon_sess._coordinated_creator.tf_sess.run(
                training_model.iter_handles[target])
            # Init all target datasets.
            mon_sess._coordinated_creator.tf_sess.run(training_model.iter_init_ops[target])

```
Instead you should use session run hooks to make it proper:
```
    class MyInitializerHooks(tf.train.SessionRunHook):
      def after_create_session(self, session, coord):
        for target in targets:
            # Get all target datasets handles.
            handle[target] = session.run(training_model.iter_handles[target])
            # Init all target datasets.
            session.run(training_model.iter_init_ops[target])
```

**Comment by MtDersvan** [comment]
@mrry I shared the graph just to illustrate that `MonitoredTrainingSession` is initialized repeatedly during a training. The graph looks that way because the checkpoint is being saved rarely. The problem is that there are no errors whatsoever. 
In the beginning I thought it maybe restarting due to OOMs and just not showing the errors, but I checked with bigger models and tf catches OOMs just fine.
My setup is an alpha k8s cluster on GCP that I used successfully since it's release this summer:
```
# Cluster settings.
CLUSTER_NAME=#####
NUM_NODES=10
ZONE=us-west1-b
PROJECT=#####
CLUSTER_VERSION=1.7.8-gke.0
DISK_SIZE=75
IMAGE_TYPE=COS
MACHINE_TYPE=n1-highmem-4
TYPE=nvidia-tesla-k80
COUNT=1

# NodePool Settings.
NP_NAME=ps-cpu
NP_NUM_NODES=1
NP_MACHINE_TYPE=n1-highmem-2

sudo gcloud alpha container clusters create ${CLUSTER_NAME} \
    --num-nodes=${NUM_NODES} \
    --zone=${ZONE} \
    --project=${PROJECT} \
    --cluster-version=${CLUSTER_VERSION} \
    --disk-size=${DISK_SIZE} \
    --image-type=${IMAGE_TYPE} \
    --machine-type=${MACHINE_TYPE} \
    --accelerator=type=${TYPE},count=${COUNT} \
    --enable-kubernetes-alpha \
    --scopes storage-full

sudo gcloud alpha config set container/cluster ${CLUSTER_NAME}

sudo gcloud alpha container clusters get-credentials ${CLUSTER_NAME} \
    --zone=${ZONE}

# This is a hack from github.com/ContainerEngine/accelerators/tree/master/cos-nvidia-gpu-installer
# - the only possible way to expose GPUs of GKE nodes for now.
sudo kubectl create -f https://raw.githubusercontent.com/ContainerEngine/accelerators/master/cos-nvidia-gpu-installer/daemonset.yaml

sudo gcloud alpha container node-pools create ${NP_NAME} \
    --num-nodes=${NP_NUM_NODES} \
    --zone=${ZONE} \
    --disk-size=${DISK_SIZE} \
    --image-type=${IMAGE_TYPE} \
    --machine-type=${NP_MACHINE_TYPE} \
    --scopes storage-full \
    --cluster=${CLUSTER_NAME}
```
I have 1 chief, 8 training workers and 1 evaluation worker.
Everything is being controlled by `MonitoredTrainingSession` in a data-parallel way.

@ispirmustafa As I mentioned before, I tried this fix https://github.com/tensorflow/tensorflow/issues/12859#issuecomment-327783827, instead of `
mon_sess._coordinated_creator.tf_sess.run()` and still got the same error. The latter is just for multi-task iterator initialization convenience. I haven't tried it again after @mrry's fix with `shared_name`, but I assume it's not the cause, as they both are due to session restarts. Though, I am going to add a proper `SessionRunHook` as you suggested, to eliminate to possibility of an internal leak.

My biggest confusion is that there is no INFO on to why `MonitoredTrainingSession` restarts without warnings.

**Comment by ispirmustafa** [comment]
@isaprykin do you have some idea on what's going on?

**Comment by MtDersvan** [comment]
@ispirmustafa I can confirm now that using [InitializerHook](https://github.com/tensorflow/tensorflow/issues/14613#issuecomment-347719771) does not fix the behaviour.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by isaprykin** [comment]
@MtDersvan is this still an issue with more recent releases?  Are there log messages from coordinator?  From saver that it is "Restoring parameters"?

**Comment by MtDersvan** [comment]
@isaprykin I am at the stage of deploying these models rn, but I'll get back to training a new batch in the following 1-2 weeks. I will then try an r1.5 and a new k8s version, and give a comprehensive feedback if something goes wrong. 
So you can either close this issue and I will reopen/open it if it's reproducible, or wait till I have the feedback.

**Comment by cy89** [comment]
Thanks, @MtDersvan. We'll leave at "awaiting response"; let us know what you find.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by MtDersvan** [comment]
@cy89 @isaprykin 
I think it's not an Iterator error.
I substituted `from_string_handle` for `make_initializable_iterator`, but the error (silent session restarts without any failures) persisted during `MonitoredTrainingSession`, so I kept debugging a bit deeper.

I explicitly started to check for exceptions in session run calls [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/client/session.py#L912), and got an uncaught `OS Error`, that was causing silent restarts:

~~~
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1392, in _do_call
    return fn(*args)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1371, in _run_fn
    target_list, status, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/errors_impl.py", line 516, in __exit__
    c_api.TF_GetCode(self.status.status))
tensorflow.python.framework.errors_impl.UnavailableError: OS Error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 910, in run
    run_metadata_ptr)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1165, in _run
    feed_dict_tensor, options, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1386, in _do_run
    options, run_metadata)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/client/session.py", line 1405, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.UnavailableError: OS Error
~~~

I could only trace it till `pywrap_tensorflow.TF_PRun()`. It's not clear what is wrong and what to do about it.
Is this an intended behaviour?  :man_shrugging:
Is there any way to avoid this it?


Env Info:
~~~
TF DOCKER: tensorflow/tensorflow:latest-gpu-py3
K8S CLUSTER_VERSION=1.9.4-gke.1
~~~

**Comment by reedwm** [comment]
/CC @isaprykin can you take a look?

**Comment by isaprykin** [comment]
Ah, it's not an OSError, it's an UnavailableError with the "OS Error" error message, right?
UnavailableError is part of MonitoredSession's _PREEMPTION_ERRORS -- the list of exceptions that it tries to recover from.  But every time it happens an loggign message should be generated.

Could you please include a more complete set of logs around the point where the session is restarted?

**Comment by MtDersvan** [comment]
Yes, it's an UnavailableError with the "OS Error" message. (sorry for a late response)
Actually, that's the problem the `run()` in `_RecoverableSession` doesn't log this:
~~~python
    logging.info('An error was raised. This may be due to a preemption in '
                 'a connected worker or parameter server. The current '
                 'session will be closed and a new session will be '
                 'created. Error: %s', e)
~~~
That's exactly what I did to debug the problem - explicitly checked for any exceptions narrowing down the scope.
All the tensorflow logs boil down to:
~~~console
...
I tensorflow/core/distributed_runtime/master_session.cc:1024] Start master session 6d9f53ba9c6c58c9 with config: allow_soft_placement: true
...
~~~

And I'm pretty sure that all the distributed workers were online. Maybe a connection issue? Also, adding a `load_balancer` reduced the frequency [but I can't say 100%].

For now I just made a custom `_RecoverableSession` that handles `UnavailableError: OS Error` without restarting the whole session (just waits a second before rerunning).
